### PR TITLE
Fix osrm-contract, tests, on Windows

### PR DESCRIPTION
**Pull request by mjjbell**
# Issue

As part of graph contraction, node renumbering leads to in-place permuting of graph state, including boolean vector elements.

`std::vector<bool>` returns proxy objects when referencing individual bits. To correctly swap boolean elements using MSVC, we need to explicitly apply `std::vector<bool>::swap`.

Making this change fixes `osrm-contract` on Windows.

We also correct failing tests and other undefined behaviours (mainly iterator access outside boundaries) highlighted by MSVC.

## Tasklist

 - [x] CHANGELOG.md entry ([How to write a changelog entry](http://keepachangelog.com/en/1.0.0/#how))
 - [x] add tests (see [testing documentation](https://github.com/Project-OSRM/osrm-backend/blob/master/docs/testing.md)
 - [ ] review
 - [ ] adjust for comments
 - [ ] cherry pick to release branch

## Requirements / Relations

 #5872 has related discussion on `osrm-contract` failing on Windows

### CODE REVIEW

File: include/server/api/base_parameters_grammar.hpp [hunk 1]
```diff
@@ -25,18 +25,20 @@ namespace
...
     template <typename Iterator> static bool parse_dot(Iterator &first, Iterator const &last)
     {
-        if (first == last || *first != '.')
+        auto diff = std::distance(first, last);
+        if (diff <= 0 || *first != '.')
             return false;
 
         static const constexpr char fmt[sizeof...(Fmt)] = {Fmt...};
 
-        if (first + sizeof(fmt) < last && std::equal(fmt, fmt + sizeof(fmt), first + 1u))
```
Comment (mjjbell) [inline include/server/api/base_parameters_grammar.hpp:39]: `first + sizeof(fmt)` creates iterator beyond end.

File: include/util/indexed_data.hpp [hunk 2]
```diff
@@ -346,7 +346,7 @@ template <typename GroupBlockPolicy, storage::Ownership Ownership> struct Indexe
         const GroupBlockPolicy block;
         block.ReadRefrencedBlock(blocks[block_idx], internal_idx, first, last);
 
-        return adapt(&*first, &*last);
```
Comment (mjjbell) [inline include/util/indexed_data.hpp:349]: For empty container, `*first` and `*last` can both be dereferencing the end iterator.

File: unit_tests/common/range_tools.hpp [hunk 3]
```diff
@@ -11,9 +11,11 @@
         const auto &rhs = {__VA_ARGS__};                                                           \
         BOOST_CHECK_EQUAL_COLLECTIONS(lhs.begin(), lhs.end(), rhs.begin(), rhs.end());             \
     } while (0)
-#define CHECK_EQUAL_COLLECTIONS(lhs, rhs)                                                          \
```
Comment (mjjbell) [inline unit_tests/common/range_tools.hpp:14]: Failed if `lhs` or `rhs` is expression creating a value (e.g. a literal) - `begin()` and `end()` would reference different containers.

File: unit_tests/util/indexed_data.cpp [hunk 4]
```diff
@@ -38,8 +38,8 @@ void test_rw(const Offsets &offsets, const Data &data)
 
     for (std::size_t index = 0; index < offsets.size() - 1; ++index)
     {
-        typename IndexedData::ResultType expected_result(&data[offsets[index]],
-                                                         &data[offsets[index + 1]]);
```
Comment (mjjbell) [inline unit_tests/util/indexed_data.cpp:42]: Potential subscript out of range.

File: unit_tests/partitioner/multi_level_partition.cpp [hunk 5]
```diff
@@ -175,4 +178,56 @@ BOOST_AUTO_TEST_CASE(mlp_sorted)
     BOOST_CHECK_EQUAL(mlp.EndChildren(4, 0), 2);
 }
 
+BOOST_AUTO_TEST_CASE(large_cell_number)
```
Comment (mjjbell) [inline unit_tests/partitioner/multi_level_partition.cpp:181]: Used a cruder version of this to detect #5878 
Adding to prevent regression.

File: appveyor-build.bat [hunk 6]
```diff
@@ -143,34 +147,41 @@ ECHO running server-tests.exe ...
...
-::XCOPY %test_region%.osrm mld\
-::%Configuration%\osrm-contract.exe %test_region_ch%.osrm
-::%Configuration%\osrm-contract.exe --core 0.8 %test_region_corech%.osrm
-::%Configuration%\osrm-partition.exe %test_region_mld%.osrm
-::%Configuration%\osrm-customize.exe %test_region_mld%.osrm
-::XCOPY /Y ch\*.* ..\test\data\ch\
-::XCOPY /Y corech\*.* ..\test\data\corech\
-::XCOPY /Y mld\*.* ..\test\data\mld\
-::unit_tests\%Configuration%\library-tests.exe
+ECHO running partitioner-tests.exe ...
```
Comment (akashihi) [inline appveyor-build.bat:150]: :+1:

**Review by akashihi — APPROVED** [review]
